// vite-plugin-ssr/dist/cjs/node/plugin/plugins/packageJsonFile.js
"use strict";
/*
 * We create a file `dist/server/package.json` to support ESM users.
 * Otherwise, following error is thrown:
 *   Must use import to load ES Module: dist/server/pageFiles.js
 *   require() of ES modules is not supported.
 *   require() of dist/server/pageFiles.js from node_modules/vite-plugin-ssr/dist/esm/node/runtime/page-files/setup.js is an ES module file as it is a .js file whose nearest parent package.json contains "type": "module" which defines all .js files in that package scope as ES modules.
 * Reproduction: https://github.com/brillout/vite-plugin-ssr-server-import-syntax
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.packageJsonFile = void 0;
const rollupIsEsm_js_1 = require("../shared/rollupIsEsm.js");
const utils_js_1 = require("../utils.js");
function packageJsonFile() {
    let config;
    return {
        name: 'vite-plugin-ssr:packageJsonFile',
        apply: 'build',
        configResolved(config_) {
            config = config_;
        },
        generateBundle(options, bundle) {
            if (!(0, utils_js_1.viteIsSSR)(config))
                return;
            const isEsm = (0, rollupIsEsm_js_1.rollupIsEsm)(options);
            const fileName = 'package.json';
            if (bundle[fileName])
                return; // E.g. already generated by Telefunc / vite-plugin-ssr
            this.emitFile({
                fileName,
                type: 'asset',
                source: getPackageJsonContent(isEsm)
            });
        }
    };
}
exports.packageJsonFile = packageJsonFile;
function getPackageJsonContent(isEsm) {
    if (isEsm) {
        return `{ "type": "module" }`;
    }
    else {
        return `{ "type": "commonjs" }`;
    }
}
