// sats-connect/dist/index.mjs
// src/index.ts
import {
  BaseAdapter,
  RpcErrorCode,
  defaultAdapters,
  getDefaultProvider,
  getSupportedWallets,
  removeDefaultProvider,
  setDefaultProvider
} from "@sats-connect/core";
import { makeDefaultConfig } from "@sats-connect/make-default-provider-config";
import {
  close,
  loadSelector,
  selectWalletProvider,
  walletClose,
  walletOpen
} from "@sats-connect/ui";
export * from "@sats-connect/core";
var Wallet = class {
  providerId;
  defaultAdapters = defaultAdapters;
  createCustomConfig;
  isProviderSet() {
    return !!this.providerId;
  }
  setCreateCustomConfig(createCustomConfig) {
    this.createCustomConfig = createCustomConfig;
  }
  async selectProvider() {
    const providers = getSupportedWallets();
    if (providers.length === 0) {
      throw new Error("No wallets detected, may want to prompt user to install a wallet.");
    }
    const selectorConfig = this.createCustomConfig ? this.createCustomConfig(providers) : makeDefaultConfig(providers);
    const nextProviderId = await selectWalletProvider(selectorConfig);
    this.providerId = nextProviderId;
  }
  async disconnect() {
    await this.request("wallet_renouncePermissions", void 0);
    this.providerId = void 0;
    removeDefaultProvider();
  }
  async request(method, params) {
    loadSelector();
    const defaultProvider = getDefaultProvider();
    if (!this.isProviderSet()) {
      if (defaultProvider) {
        this.providerId = defaultProvider;
      } else {
        try {
          await this.selectProvider();
        } catch {
          return {
            status: "error",
            error: {
              code: RpcErrorCode.INTERNAL_ERROR,
              message: "Failed to select the provider. User may have cancelled the selection prompt."
            }
          };
        }
      }
    }
    const adapter = this.defaultAdapters[this.providerId];
    walletOpen(this.providerId);
    const response = adapter ? await new adapter().request(method, params) : await new BaseAdapter(this.providerId).request(method, params);
    walletClose();
    if (response?.status === "error" && response.error?.code === RpcErrorCode.USER_REJECTION) {
      if (!defaultProvider) {
        this.providerId = void 0;
      }
    } else {
      setDefaultProvider(this.providerId);
    }
    close();
    if (!response) {
      return {
        status: "error",
        error: {
          code: RpcErrorCode.INTERNAL_ERROR,
          message: "Wallet Error processing the request"
        }
      };
    }
    return response;
  }
  addListener = (event, cb) => {
    const defaultProvider = getDefaultProvider();
    if (!this.isProviderSet() && defaultProvider) {
      this.providerId = defaultProvider;
    }
    if (!this.isProviderSet()) {
      throw new Error(
        "No wallet provider selected. The user must first select a wallet before adding listeners to wallet events."
      );
    }
    const adapter = this.defaultAdapters[this.providerId];
    if (!new adapter().addListener) {
      console.error(
        `The wallet provider you are using does not support the addListener method. Please update your wallet provider.`
      );
      return () => {
      };
    }
    return new adapter().addListener(event, cb);
  };
};
var src_default = new Wallet();
export {
  src_default as default
};
